### ISSUE TITLE
Implement support for Database Unit Tests

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#3110]
# TL;DR

I think adding database-driven unit tests is extremely viable with only 1 small change to existing Rock core code necessary.

# Feature Request

@nairdo pointed me towards a project called Effort (http://entityframework-effort.net/) that I took a look at. It is extremely promising and I think would work to get us 98% of the way there for unit testing against a database - without having to keep a somewhat complex setup of having a SQL Server database that must be kept in a known good state.

## Existing Problems

Currently, in order to run entity tests you must have a database. This presents problems both for automated testing as well as manual testing.

With manual testing, it requires that you have a SQL Server database populated with the bare minimum of required sample data. It also requires that each test not make any changes to the database. This can be difficult since the Rock internals will sometimes automatically create data in the background on a new RockContext you don't have control over (such as EntityTypes). This means potentially having to run tests one at a time and rolling back changes to the database manually (or via some script you run).

With automatic testing, you have the same issues but since it's automatic a real SQL Server database is probably not available, even if we could overcome the other problems.

## Potential Solution

The Effort project does two things for us.

First, it uses an in-memory database so that no external database is needed. This is essentially a SQL Server database stored in RAM. It does _not_ use an actual SQL Server, thus it is probably not as optimized as it could be - but for unit testing that is probably okay.

Second, it allows us to load in sample data via CSV files. They have a tool for generating the CSV files. I had to make a small tweak to so that it would work with a Rock database. I believe this change was simply adding a reference to the Spatial data types DLL. When the in-memory database is created, it is pre-loaded with data from the CSV files. I did a test-export from a newly installed 8.0 system and used that data and it worked fine.

So, this leaves us at a spot where we should be able to use this for both manual testing (i.e. in Visual Studio clicking the Run Tests button) as well as automated build testing (i.e. CI / appveyor).

## Drawbacks

There are a few drawbacks to be aware of, but I think they are fairly small when they allow us to potentially test so much of the data layer.

### Geospatial columns

These are not supported. Remember, this is not a true Microsoft SQL Server. There is a small chunk of code that must be added to the RockContext.OnModelCreating() method. It checks if the Database Connection Provider is the Effort one (i.e. we are in unit test mode) and if so it marks all the Geospatial columns as ignored. If new Geospatial columns are added, this would need to be updated. On the plus side, you would know instantly because all tests would fail to run.

For the record, other than the Analytics tables, the only core table affected is the Location table.

### CSV files are annoying

Keeping 109 CSV files around is kind of annoying. Not to mention it takes up about 50MB (this is a bare system, the sample data has not been loaded).

The Effort system is fairly well componentized. So I built a new data loader class that pulls the CSV data from files inside a ZIP file (i.e. select all CSV files, Send To Zip, and then use that zip file). This brings the file count down to 1 thing to keep track of. And it reduces the file size to 15MB. For the record, there was no noticeable increase in time from using a zip file vs straight CSV files.

### Tests are a bit slow

Speaking of speed, lets talk about the test speeds. Remember, we are populating a database at the start of every single test. Here is the configuration I am running these tests on:

* Virtualized Windows 10 inside VirtualBox (running on Windows 10) - 64-bit.
* i5-8400 @ 2.8Ghz x 2 cores
* 6GB RAM
* SSD via USB for C: drive
* Visual Studio 2017
* Tests do very simple queries, such as `personService.Queryable().Count()` so nearly all of the timing is overhead.

(Note: I mention threads next. Sometimes the XUnit system runs the tests in parallel, sometimes it doesn't. I haven't figured out why it's random)

The first test to run on each thread takes 40 seconds to run.  Each follow up test on the thread takes 9 seconds to run. This is due to Effort caching the "original state" of the database so when the next test requests the database be reset, it doesn't need to do as much work.

So, in my test class I have 6 unit tests. When running single-threaded it takes 85 seconds (9 * 5 tests; 40 * 1 test). When running in parallel at 2 threads it takes 58 seconds (9 * 2 tests; 40 * 1 test; over 2 threads).

### Database Computed Columns

Since there is no database, I don't think these properties will be correctly populated. I think most of these computed columns are just "nice to have" values. For example, the `Person.BirthDate` is computed based on the individual BirthYear, BirthMonth and BirthDay columns. Because we provide a `get` method for this property, if we access `person.BirthDate` we should get a valid value for unit testing. If we try to do something like `personService.Queryable().Where( p => p.BirthDate >= DateTime.Now )` we will probably get bogus results.

Given the fact that most(all?) database generated columns provide an in-memory getter to calculate the value, I would consider this a minimal issue. It's just something that we need to be aware of when writing unit tests.

## Potential Issues

Okay, so there are a few issues we would need to overcome. Nothing daunting, just some design decisions. My goal for this is to not only allow for Rock to have internal database/Entity tests, but also for my plugins (and others) to be able to build database/Entity unit tests into their plugins. Two good examples that I'm sure would benefit from this is the Room Management and the Project Management plugins. They are both fairly complex and I'm sure neither of us wants to make a "minor change" to the code and break some seemingly unrelated thing. With that in mind...

### New DLL

There is some code that handles the loading of the Zip files as well as handles the mock database initialization. I think these should be moved into a new DLL. `Rock.TestSupport` or something like that. This way a plugin could include Rock.Dll and Rock.TestSupport.Dll into their own test project and not have to worry about it running the entire suite of Rock tests too when they click the "run all tests" button.

We probably wouldn't want to include this DLL in the main RockWeb/bin folder, but as long as a pre-compiled version of it was available (maybe on the Releases tab on github) that should be fine.

We _could_ include it with the Rockit SDK download, but we also need to include the Effort and NMemory DLLs since they would be dependencies.

### Split out existing tests

Because of the speed issue, it might be worth splitting out the need-database tests from the logic-only tests. In the Rock solution, the Run All would still run both, but this would make it easier to run them separately if we need to. Or, for example, we might want the CI system to only run the non-database tests (again because of the time required) and a core developer runs the full-suite every once in a while and before a release is tagged.

I'm not sure how critical this one is. Maybe it doesn't really matter that much, but a point for consideration.

### Database Changes

I'm not sure how often we will need to re-populate the "basic" CSV data files. Meaning, if a new column is added I don't know if we instantly need to generate new CSV files. Does column order matter when it imports the data?

If even minor changes cause issues then we might want to look at customizing the CSV importer to better handle that so we don't have to re-generate for minor changes - such as adding a new INT column with a default of 0 (zero). i.e. we don't need to set that, in memory it will be treated that way anyway.

### Generating CSV files

Effort includes a tool for this. It works very well, but is a bit cumbersome as it's meant to be VERY generic. We may want to fork this and do a few tweaks to simplify it's operation. Even just having it remember the last settings used would be great.

For the record, once I understood how the tool worked, it took me about 2 minutes to generate the CSV files from an existing database. Coupled with the RockDevBooster tool, generating a clean 1.8.0 database and exporting it took about 9 minutes (not including download time from github).

### Referencing CSV files

I'm not sure where this ZIP of CSV files should go. There are a few options I can think of:

1. Hard code the path to the Solution path, e.g. `[LoadData( @"../../../RockDataBase.zip" )]` and make it up to the developer to put a file there if they want those tests to work.
2. Store those files somewhere centrally (this could get trouble some as people work on different features, but it's still possible to do based on Rock version or something): `[LoadData( @"http://rockrms.com/RockDataBase-{RockVersion}.zip"]`
3. Embedded in Git somewhere (I don't really like this solution, the git repo is big enough as is) and built into the test DLL as a resource: `[LoadData(@"res:RockDataBase.zip")]`

I would also consider this a minor issue. Meaning, I'm sure we can come up with a way that makes sense and doesn't put a lot of undue load on the dev to keep yet another thing up to date and in sync.

# Current State

As mentioned above, I have some very simple unit tests working. I have the Zip data loader working. Though most of the custom code still needs to be documented.

Below is a sample of a minimum unit test class for reference:

```
namespace Rock.Tests.Rock
{
    [LoadData( @"C:\Users\Daniel Hazelbaker\Desktop\RockDataBase.zip" )]
    public class MoqTest : DatabaseFixture
    {
        [Fact]
        public void MyTest()
        {
            using ( var rockContext = new RockContext() )
            {
                var adults = new PersonService( rockContext ).GetAllAdults();
                var adultCount = adults.Where( p => p.NickName == "Giver" ).Count();

                Assert.Equal( 1, adultCount );
            }
        }
    }
}
```

The `DatabaseFixture` does the setup of the database before each test. Basically it's a helper tool so the test writer doesn't need to think about any of that fancy stuff.

The above test is pretty pointless as we are just verifying the data already in the database. But suppose we want to test that the auto-matching logic hasn't broken. We could have a few unit tests that use the existing data to verify that a match on Ted Decker with his e-mail address finds the existing record rather than creating a new one (or in the case of some test, that it DOES create a new record).

# Final Thoughts

I'm sure I've missed things and forgotten to mention things. I don't have the code committed to a branch anywhere yet, but if anybody wants to see what changes I have had to make up to this point I can commit to a junk branch for review.

Please let me know if you have questions or comments to help further this as a potential inclusion into core.

**Providing Code**
I will provide code if feature request is approved: **Yes**

**Comment by edmistj** [comment]
Since @nairdo has been working on the unit testing I'll let him take this one.

**Comment by cabal95** [comment]
Updating some information on this.  I have been doing some more testing on this since discovering that AppVeyor actually supports full SQL server unit testing. So I have been able to do some massive simplification.

* Effort framework is no longer used.
* No changes to any existing Rock.dll code is required.
* Geospatial columns work.
* Database computed columns work.

Basically, the way I am going about this now with the simplification, is to essentially build a "bare bones" Rock database. (Optionally shrink the database files as much as possible). Detach the database and then zip up the MDF and LDF files (this created a roughly 30MB file for me).

Then, in the app.config file, we add an AppSetting that indicates the location of this file (for AppVeyor, this should probably be a URL to the latest version of the Zip file).

Xunit provides an `IClassFixture` that allows for shared initialization and de-initialization of all tests in the test class. Meaning, if you have 15 tests in the class, the init/de-init code is only run once.

We have a custom `DatabaseFixture` class that will take the zip file, extract the MDF and LDF files and then attach the database to the full blown SQL Server, as specified in the app.config connection string.

So, on the assumption that we have to test classes, each with various tests, we would have the following process:

1. Begin testing Group A tests.
2. Extract and attach database files (also clear Rock cache).
3. Run all 10 tests in Group A.
4. Delete database created in step 2.
5. Begin testing Group B tests.
6. Extract and attach database files (also clear Rock cache).
7. Run all 18 tests in Group B.
8. Delete database created in step 6.

Performance is massively improved. Running two groups of tests, one with 3 tests and the other with 4, takes a total of 47 seconds on my system. That is extracting zip and attaching database two times. A good chunk of that 47 seconds is spent on doing the one-time DLL loading for EF. Each database load seems to take only about 5 seconds and the individual tests are in the `ms` range.

**Comment by nairdo** [comment]
@cabal95 some of this sounds like where we've been going with the new https://github.com/SparkDevNetwork/Rock/tree/develop/Rock.Tests.Integration project (MSTest based vs XUnit).  Looks like the performance is on par with yours (some of these tests attempt 1M attendance code operations which is why they take many seconds):

![image](https://user-images.githubusercontent.com/383807/43725643-f3b291d4-9951-11e8-828a-4cd87bdf49c0.png)


I liked the idea of having the app.config [refer to a app.ConnectionStrings.config](https://github.com/SparkDevNetwork/Rock/blob/develop/Rock.Tests.Integration/app.config#L42) that can be deployed/created in the running environment, but I could be persuaded to use a static value instead. 

Also, having a method that can restore that db back to its initial state (in either [[ClassInitialize] or [TestInitialize]](https://github.com/SparkDevNetwork/Rock/tree/develop/Rock.Tests.Integration#mstest-lifecycle)) quickly is a big plus.

One issue I've run into is needing to keep the database migrated -- so we might want to tackle that as well.

Let me know if you'd like to merge our two efforts into one.

**Comment by cabal95** [comment]
I'll work on pushing my code to a public branch so you can take a look. It's basically contained in a single class and should easily work with either MSTest or XUnit (or NUnit for that matter).

Basically, the concept is that before each test runs, the existing database is forcibly deleted (this closes existing RockContext connections) and then re-stored back from the data in the ZIP file. So even if a person had to implement this at the start of each individual test, it's a single method call (ResetDB). The one thing this means is that database tests CANNOT run in Parallel since they would be hitting the same database - but I think the same is probably true of any method that uses an actual database.

After I get my code pushed I'll take a look at yours too and see if it makes sense to merge these together.

**Comment by cabal95** [comment]
@nairdo Sorry been a busy day. I pushed my local branch to my github repo:

* https://github.com/cabal95/Rock/blob/feature-dh-unit-tests/Rock.Tests/Data/DatabaseFixture.cs

That file contains about 98% of the logic. 

It looks like conceptually we are doing the same thing. Your method (depending on the individual test to clean up after itself) will probably see  overall better performance since you aren't restoring a 150MB database for each test.

My method is slower, but ensures the database is 100% back to clean state for each test (many tests this doesn't matter as much, but some tests will cause Rock to modify data in the background so trying to clean up gets complicated).

It would also be possible to mix them together and let a class attribute decide if the database is re-loaded before EACH individual test, or just for the entire class. And maybe that would be the best way to go.

For example, all your attendance code tests are probably safe to run using the same database and just cleanup after themselves. So maybe the entire test class loads the database from ZIP (for example) before the entire class suite of tests starts and then it is dropped after the whole suite has finished.

Other more complex cases (for example, when testing history record generation) might flag that each and every test case/method needs a clean database to work with.

I think it would make sense to see about merging these together.

Also - if we are not worried about running these tests on AppVeyor, we can set the app.ConnectionStrings.config file to be stored in git setup to use a LocalDB. This would make it "clone and run" for most developers.

I would love to end up with a final product/model that plug-in developers can use to build data-driven unit tests for their own plugins. That is what kicked this off for me. We have a Remote Check Deposit plugin being worked on that touches various parts of the system (accounts, transactions, images, global attributes, custom models, etc.). We want to have a unit test that runs an export of a file to be uploaded to the bank and compares the SHA-1 of the file with the "known good file" SHA-1 to ensure that any future changes don't _inadvertently_ change anything.

**Comment by cabal95** [comment]
Discussion closed here. New vastly simplified implementation open for discussion at #3227
